### ISSUE TITLE
VPN Blocking internet connection on iOS

### ISSUE THREAD

**Issue by birdsarah** [mozilla-mobile/mozilla-vpn-client#1265]
**Open question: is this iPad specific?**

I can describe the symptoms. iOS v2.3 has been in this state on my iPad predominantly for the last while. Unfortunately - I don't know how to get it into this state. But once it's in this state it's been staying like this for a while. To remedy the situation I tried:

* closing and opening the app (see notes below)
* turning off ipv6
* uninstalling and reinstalling

After uninstalling and reinstalling, the problem persisted for a bit then seemed to fix itself, then it came back again...I'm sorry I can't be more precise about this. I'm not confident how to replicate.

Symptoms (screenshots posted below but i'm not sure they're that helpful):

* open vpn, hit connect. button stays in "connecting" state and never connects. ios indicates that is on, but cannot connect to internet.
* close the app and reopen, app then says it's connected and the notification pops. ios still indicates that vpn is on. still cannot connect to internet. (@bakulf has explained that this makes sense because on opening the app the app just checks the ios state and given that ios says vpn is on, app just proceeds)
  * from this state if you click "i" for information the IP addresses just read as "Loading"
  * from this state you can click to switch server, but then you just go back to "connecting" stuck state

One more thing to note. On reopening the app, the app will briefly pop a red "internet not connected" warning. But, that warning then goes away by itself.

![BF0E67E9-72B1-463B-A6E7-CE3631AF1B37](https://user-images.githubusercontent.com/1796208/124678744-cbbcd200-de88-11eb-9e27-71993643a372.png)
then close app and reopen

![A4740D2F-AE6E-48D6-BDF1-2E1FC612DF51](https://user-images.githubusercontent.com/1796208/124678542-5bae4c00-de88-11eb-8c6f-f21ed5d7ca3e.png)
^^ screenshot didn't capture the notification

![357AFA64-F8F4-40D2-A1EA-C395A6156DD4](https://user-images.githubusercontent.com/1796208/124678541-5a7d1f00-de88-11eb-9416-4c5807188c77.png)

![1034CF26-43CF-48D6-B768-6D7F6CA147AB](https://user-images.githubusercontent.com/1796208/124678536-58b35b80-de88-11eb-89a4-77b840e6a11f.png)
![B8EA91C5-8613-4B4B-96BD-CB3DBE0C76B5](https://user-images.githubusercontent.com/1796208/124678545-5c46e280-de88-11eb-9125-b49577f06977.png)

The worst part about this is you have to know to kill the app to load it to, "turn the vpn off" so you can then use the internet again. I wonder if we should allow toggling off during connecting state so people don't get stuck on screenshot one

┆Issue is synchronized with this [Jira Bug](https://mozilla-hub.atlassian.net/browse/VPN-693)

**Comment by birdsarah** [comment]
I'm posting logs from "Console" to @mbirghan for him to take a look at.

**Comment by birdsarah** [comment]
We are thinking this might be iPad specific. Please comment on this issue or let support know if you're having this issue and tell us whether you're on an iPad or not.

**Comment by birdsarah** [comment]
Here's logs from my device. At one point during this set of logs the VPN did work. Then it stopped again. https://gist.github.com/birdsarah/306f8e0671c1133f36734a2e1582eabc

**Comment by jr-johnson-support** [comment]
https://drive.google.com/file/d/1xg1hK0rVMXGArZFr6e2OJJXy4Af7Dj7f/view?usp=sharing video was a bit too big to upload. Added to drive and shared file link to download and view.
[jrLogs.txt](https://github.com/mozilla-mobile/mozilla-vpn-client/files/6794018/jrLogs.txt)

**Comment by jr-johnson-support** [comment]
[JR logs2_3_3.txt](https://github.com/mozilla-mobile/mozilla-vpn-client/files/6794035/JR.logs2_3_3.txt)

https://user-images.githubusercontent.com/58608126/125138664-a7edcc00-e0d4-11eb-9fd8-f43287134a0c.MP4

realized my ipad version was 2.3.0 in the comment above. updated to 2.3.2 and replicated the issue.

**Comment by birdsarah** [comment]
I may have come up with a way to reproduce, am verifying with @mbirghan.

**Comment by birdsarah** [comment]
We finally have a reproducible test case. The issue is linked to trying to connect to the VPN when another app is trying to detect if you're on a VPN.

e.g. 

1) Open hulu, netflix, etc and start playing a show
2) Open VPN and try to connect

Steps 1 and 2 need to be done fairly quickly together for some apps that only check the vpn status at the start of playing media (e.g. peacock). Others (e.g. hulu) are always checking if you're on a VPN.

In this scenario, the VPN will exhibit the symptoms described originally.

**Comment by data-sync-user** [comment]
➤ Nma-Sie-Anyene Davenport commented:

Per Moritz:    https://docs.google.com/document/d/1HZE5EKYLqqpx4EqkiVuvmV95OP5d5lZxm1OzedCgOD0/edit ( https://docs.google.com/document/d/1HZE5EKYLqqpx4EqkiVuvmV95OP5d5lZxm1OzedCgOD0/edit|smart-link )

**Comment by data-sync-user** [comment]
➤ Sarah Bird commented:

Adrienne Davenport I’d like to see this be a priority for v2.6, I still can’t use my production subscrription of my VPN on my ipad.

**Comment by data-sync-user** [comment]
➤ Sarah Bird commented:

Just confirmed that this is still an issue in production v2.5 VPN on my iPad.

**Comment by data-sync-user** [comment]
➤ Moritz Birghan commented:

After looking into the logs, I found that every time the issue is happening, this line can be found in the backend logs multiple times: Failed to receive packet: address family not supported by protocol family.

After some more research, I found that this is caused by a call of ReadFromUDP by wireguard-go. This function interacts with the OS and can fail.

Even though I cannot be 100% sure that this is the cause of the issue and not just a symptom, everything points to this. Sarah Bird and I were able to find out that it is only an issue on older iOS versions and not on newer ones (iOS 15 and later). Since this issue seems to have OS involvement and ReadFromUDP does interact with the OS, it points to ReadFromUDP being the issue. The symptom of not being able to connect to the internet can also be explained since if wireguard-go cannot read packets, it would not allow any connection.

Currently, I am only aware of staff members encountering the issue, so it does not appear to be an issue in the wild. Also, since this is an issue inside of wireguard-go, we can do nothing to fix it. We already have a mitigation in place to allow users to recover from the broken state. 

I will close this issue, update wireguard-go and reach out to support to be aware of it and notify me if users are experiencing it.
